### DISCUSSION TITLE
How to re-style the EM tag

### DISCUSSION THREAD

**Discussion by jsbham** [lwindolf/liferea#1201] [Q&A]
Liferea 1.13.7
libwebkitgtk-4.0-37
Ubuntu 22.04.1 LTS

The ArsTechnica feed uses the EM tag to enclose certain text, and it renders as green, which is difficult to read. I'm not sure where this is coming from, as I can't find it in any preferences using dconf-editor or mate-control-center.
![ars_content](https://user-images.githubusercontent.com/123129367/213923197-a4530f32-3f5d-443e-bbd3-dcfebc884848.png)

I attempted to remedy this by placing the following in liferea.css
```
em {
  color: #111111 !important;
  font-style: normal;
  font-weight: bold;
}
```
I tried to be sure there'd be no caching issues, by exiting liferea, ensuring no webkit processes were still running, and removing all webkit cache blobs.

This did not fix the problem. I note that other stylings I have in liferea.css do take effect, for example BODY, so it doesn't seem as if I would need any further specification for the EM style.

I have the following settings, in case it matters:
 - Browser: Never run external Javascript
 - Privacy: Use Reader mode

Also, in case it matters, I run Fvwm as my window manager, and I don't run mate-session or gnome-session. I do start mate-settings-daemon in my fvwm.rc.

My assumption, at this point, is that default web styling for text is coming from GTK somewhow, but I thought that using liferea.css would be the best way to override this.

Any help appreciated.

**Comment by lwindolf** [comment]
I tried to reproduce it, but cannot at the moment. The "Enlarge" image caption currently does not contain `<em>` tags. They probably changed there layout.

To answer your question in such a case start Liferea with CLI param `--debug-html` and from the HTML view context menu choose "Inspect" to start the Webkit Inspector. Ensure to open it in a separate window and you should be able to debug the issue.

**Reply by jsbham** [reply]
Thanks, Lars. Very interesting. I guess I had mis-read the feed text, because the EM isn't used where I'm seeing the green text. What is happening is that for some reason, if there is an ANCHOR (no HREF) tag, it is causing the text to be green. The following LINK tag terminates this.
This from the Atom file:
```
<div><a name="page-1"></a></div>
  <ul class="gallery shortcode-gallery gallery-wide">
          <li>
        <div class="gallery-item-content">
          <div class="gallery-image"> image tag snipped </div>
          <p>
            HP sees the Envy Move being shared among family members and moved from room to room.                          [credit:
                              HP                            ]
                      </p>
        </div>
      </li>
      </ul>

<p>In a tech world with powerful desktops and
```
gets transformed into this HTML
```
<div id="readability-page-1" class="page"><div name="page-1">
  <ul>
          <li>
        <div>
          <p> image tag snipped </p>
          <p>
            HP sees the Envy Move being shared among family members and moved from room to room.                          [credit:
                              HP                            ]
                      </p>
        </div>
      </li>
      </ul>

</div><p><a name="page-1">In a tech world with powerful desktops and</a>
```
The _a name="page-1"_ has gotten moved so that it's now after the P tag for the first paragraph of text.

It might take some digging to find out why that transform is happening.

This happens only when the P has an A following it. For a feed item that has:
```
<div><a name="page-1"></a></div>
<p>A jury this week awarded $228.5
```
The HTML matches the ATOM feed content.

I hope I have explained this well enough. Unfortunately, since it's using an A tag, I don't know that I can style it. It seems a bug to me that the HTML engine would restyle text following an A tag which is only an anchor, and not a link. For a test, I made a simple HTML file and loaded it into Firefox with only this:
```
<BODY>
<p><a name="test this color">This is text which should be black<a href="/localhost/">localhost</a>
</p>
</body>
```
And the test text received no styling.

I have another case, just found this one. I had thought maybe the issue was with reader mode, however:
```
<div dir="ltr">
        <div id="content" class="content"><div id="readability-page-1" class="page"><div name="page-1">
<p>A jury this week awarded $228.5 million to seven plaintiffs in their case against Nevada-based water company Real Water, which sold alkaline water tainted with hydrazine, a highly toxic chemical found in fuel for rockets and spacecraft.</p>
<p>The plaintiffs included a 7-month-old boy who was hospitalized with severe liver failure and nearly needed a liver transplant. Another was a 69-year-old woman who was hospitalized for liver failure after drinking the water for years. She died in the hospital on November 11, 2020.</p>
<p>They are among around a hundred people who say they have suffered acute liver toxicity and failure from the water and have filed lawsuits against Real Water, according to attorney Will Kemp. In an interview Friday, Kemp told Ars he represents a little over 60 of the plaintiffs, including the seven in the case that received a jury verdict this week. It was the first case to go to trial.</p></div></div></div>
        <div id="item_comments" class="item_comments">
          <p>
          </p><p>
        </p></div>
      </div>
```
With reader mode off, this text is all green - there is an anchor tag preceding it, and with reader mode on, it is all black. I have found that in this case, reader mode stripped the ANCHOR tag. When there is green text, the presence of an A LINK tag terminates the green styling - the rest of the paragraph, after the link, is plain. Reader mode does not always strip the ANCHOR tag, and it seems that it also moves it, probably due to stripping out the elements in between it and the paragraph.

Maybe this is not a Liferea problem - something in WebKit?. Possibly, I can write a filter to remove any ANCHOR tags which have only a name, and no href. Or, I can try styling the A with no href, since CSS has that. However, it still seems a bug somewhere, because I think the ANCHOR should not cause that to happen.

**Reply by lwindolf** [reply]
I'm wondering if this might be happening because the anchor tag is an unclosed tag. Liferea does process all content to XHTML content using fuzzy libxml2 parsing. Something strange might be happening to unclosed anchor tags.

**Reply by jsbham** [reply]
I just noticed a typo in my response, so, for accuracy:
`in a tech world with powerful desktops and</a>`
there should not be a closing A tag. I guess I fumbled a bit in editing. What followed that was a complete A tag with href, text, and a closing tag. But, I think we do still have the same understanding - to the extent we understand what's happening.

I'm not sure what to expect for an unclosed anchor tag. In the past, the expectation was that browsers (thus HTML parsers) would be tolerant of those, though perhaps that varies depending on the doctype declared. I would expect generated XHTML to be well formed. Thus, if an A tag lacked a closing tag, and also had no text content, maybe this would be legal.
`<a name="page 1" />`
But that's not what we're seeing. So, I think webkit might just be behaving poorly, or perhaps that's the expected behavior. Maybe webkit has a default of green for anchor text, or uses a system default there.It's clear from the text display for certain elements in, for example, mate-control-center, that the green is defined somewhere, for something. I'm having difficulty finding where it's defined.

I did just discover that the "name" attribute has been deprecated, in favor of "id". So, I'm a little curious why libxml2 (if that's where it's happening) would generate an anchor with a name attribute. I'm also puzzled as to why the P tag with the name attribute is being converted to an anchor at all. I can't think of any reason to do so.

I will conclude this isn't an issue with Liferea at all. Possibly, I can point to libxml2 or some other library involved in converting the atom content to XHTML, both for generating an unclosed tag, and for changing the "p name=" to an "a name=" as either a bug, or a bad design decision. And I acknowledge that parsing possibly poorly formed markup of any sort is a challenge.

Thanks for pointing me in the right direction to find a work-around. In liferea.css:
`a:not([href]) {color: #000000 !important; }`
solves this. Note that I do have specific (not green) colors defined in liferea.css for A - link, hover, and visited.
